Display contents lists under specific conditions

emmabeynon opened the pull request:
For https://trello.com/c/2UIR4Sen/210-investigate-how-to-deal-with-short-content-lists

We have identified that there are occasions where it's not appropriate to display a contents list, for example when there is only one item on the list.  Having spiked an approach for displaying a contents list dependent on certain conditions, this PR implements the approach.  We've reached the below criteria for displaying a contents list in conjunction with Si Stephens-Manassiev in the content team:

1) We won't show a contents list that has one item.
2) We will show a contents list with 2 items if the first item's character count is over 415.
3) We will show a contents list with 2 items if the first item contains a table with more than 13 rows.
4) We will show a contents list with 2 items if the first item contains an image and character count is over 224.
5) We will show a contents list with 2 items if the first item contains an image and a table with more than 6 rows.
6) We will show a contents list with 3 items or more.

**To Do**
- [x] Update detailed guides after https://github.com/alphagov/government-frontend/pull/706 has been merged.

Component guide for this PR:
https://government-frontend-pr-719.herokuapp.com/component-guide

fofr commented on app/presenters/content_item/contents_list.rb:
```diff
@@ -22,5 +28,19 @@ def extract_headings_with_ids(html)
+    end
+
+    def first_item_content
+      element = Nokogiri::HTML(body).css('p').first
```
Some questions about this part:

* what happens if the content begins with something that's not a paragraph, eg a table or image?
* is this relatively slow at the moment? how performant could we make it?

emmabeynon replied:
Good point about the paragraphs - I'll try and find some Detailed Guides examples with other content and work something out.
Steve's got a suggestion for optimising the Nokogiri stuff below 👇

fofr commented on app/presenters/content_item/contents_list.rb:
```diff
@@ -3,6 +3,7 @@ module ContentsList
     include ActionView::Helpers::UrlHelper
     include TypographyHelper
 
+    WORD_COUNT_LIMIT = 100
```
I wonder if total word count is a useful heuristic.

steventux replied:
What @fofr said. Some formats will contain medical terminology with superlongtechnicalwords, some will contain more legalese. I think a char count might be a more honest metric.

fofr commented:
Love this as a starting point.

Are there some example URLs showing the effect?

steventux commented on app/presenters/content_item/contents_list.rb:
```diff
@@ -22,5 +28,19 @@ def extract_headings_with_ids(html)
+    end
+
+    def first_item_content
+      element = Nokogiri::HTML(body).css('p').first
```
I think we can optimise this a bit by only parsing the body once with Nokogiri, building the contents list as the existing implementation does, but also populating the first _n_ chapters so we can check their length.
This https://stackoverflow.com/questions/3484874/how-to-split-a-html-document-using-nokogiri does something similar by building a hash of chapters.

steventux commented on test/integration/topical_event_about_page_test.rb:
```diff
@@ -49,9 +49,54 @@ class TopicalEventAboutPageTest < ActionDispatch::IntegrationTest
+
+  def long_first_item_body
+    "<div class='govspeak'><h2 id='response-in-the-uk'>Item 1</h2>
+    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean facilisis
```
The [Faker gem](https://github.com/stympy/faker/blob/master/doc/lorem.md) is great for generating this sort of thing.

emmabeynon replied:
Nice one!  Will keep that in mind for production tests 👍

fofr replied:
We can also use the govuk_schema gem to generate random examples to test this with

emmabeynon commented:
> Are there some example URLs showing the effect?

We've got this as an example of the contents list not showing if there's only one item: https://government-frontend-pr-719.herokuapp.com/government/topical-events/ukraine-reform-conference/about
Unfortunately there aren't any live examples of Topical Event About Pages with two items - I've fudged some data locally to check it works - but I know that Detailed Guides has some, so I can extend this to the view for this format and will update the PR with an example 👍

steventux commented on app/views/content_items/topical_event_about_page.html.erb:
```diff
@@ -8,9 +8,15 @@
-    <%= render "components/contents-list-with-body", {
-      contents: @content_item.contents
-    } do %>
+    <% if @content_item.show_content_list? %>
```
This might be better implemented in the `contents-list-with-body` component as there's already a check for `@content_item.contents.any?`

emmabeynon replied:
Yeah definitely - I've noted it's not an ideal implementation in the PR description, and would need to investigate other formats using this component before making changes to it in prod.

emmabeynon commented:
@fofr @steventux I've added a condition for showing a contents list when a table has a certain number of rows, and a condition for when an image is present (I imagine we could use the same idea for videos).  Regarding images, I had a quick chat with Steve about this last night and we agreed that since it seems that images are always the same size, the simplest way to deal with them is to just show contents lists when they're present.

steventux commented on app/presenters/content_item/contents_list.rb:
```diff
@@ -13,14 +16,76 @@ def contents_items
+      return unless table.present?
+
+      row_count = table.css('tr').count
+      row_count > 10
```
I know this PR is still quite experimental, so a reminder to use the relevant constant `TABLE_ROW_LIMIT` here.

emmabeynon replied:
Yep, that'll be me messing around with row numbers and not replacing the constant :woman_facepalming:

steventux commented on app/presenters/content_item/contents_list.rb:
```diff
@@ -13,14 +16,76 @@ def contents_items
+      row_count > 10
+    end
+
+    def find_table
```
This looks very similar to the following method, wonder if the logic is similar enough for one method.

emmabeynon commented:
@steventux @fofr I've updated this PR to be less spikey and more production-friendly, and added some new conditions based on conversations with Si who has been testing various content lengths with a focus on common mobile displays.

steventux commented on app/presenters/content_item/contents_list.rb:
```diff
@@ -3,24 +3,112 @@ module ContentsList
+      until element.name == 'h2' do
+        if element.name == 'table'
+          return element
+        else
```
Are these `else` conditions necessary given we `return` on the `if`?

steventux commented on app/presenters/content_item/contents_list.rb:
```diff
@@ -3,24 +3,112 @@ module ContentsList
+    end
+
+    def parsed_body
+      @parsed_body ||= Nokogiri::HTML(body)
```
:+1:

maxgds commented on app/presenters/content_item/contents_list.rb:
```diff
@@ -3,24 +3,58 @@ module ContentsList
+
+    def show_contents_list?
+      return false if contents_items.count < 2
+      return true if contents_items.count > 2
```
what if contents_items.count == 2?

steventux replied:
Line 24 would catch this case but only if it's a paragraph - @emmabeynon is it ever likely to be a 2 item list which doesn't contain paragraphs?

emmabeynon replied:
Subsequent commits deal with checking for content with a lot of table rows, and also content with a mix of images, long content and long tables.

maxgds replied:
Makes sense, I hadn't looked ahead enough, sorry.

maxgds commented on test/presenters/content_item/contents_list_test.rb:
```diff
@@ -78,4 +78,45 @@ def body
+  test "#show_contents_list? returns true if number of contents items is less than 3 and the first item's character count is above 100" do
+    class << @contents_list
+      def body
+        "<h2 id='one'>One</h2>
```
I'm not sure how we can avoid this, but it feels a little fragile. If somebody updates the heading structure and all the items in govspeak become h3s then this won't happen on the live site, but the tests will continue to pass.

steventux replied:
The [existing implementation relies on the presence of `h2`s](https://github.com/alphagov/government-frontend/blob/master/app/presenters/content_item/contents_list.rb#L19), it's not a problem we're solving here. It's also a pattern we use in Govspeak frequently for anchored headings.

maxgds commented on app/presenters/content_item/contents_list.rb:
```diff
@@ -49,6 +51,23 @@ def first_item_content
+      row_count > TABLE_ROW_LIMIT
+    end
+
+    def find_table
```
I think for clarity this should be called find_first_table - I went to check whether it was scoped when I saw it called in the previous method.

steventux replied:
@maxgds it's a common convention in ruby to suffix methods which return a boolean with `?`
It's also conventional (at least with ActiveRecord) to return an object of some sort with methods prefixed with `find_`.
Given this method returns a boolean and not a table element I think the method name is OK.

maxgds commented:
Looks good overall - happy to pass this, although I would prefer that one method name be updated.
